feat: introduce constraint generation

mlutze opened the pull request:
related to #6499

magnus-madsen commented:
- Check public/private.
- Check that every def has documentation.

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+  def visitExp(exp0: KindedAst.Expr)(implicit c: TypeContext, root: KindedAst.Root, flix: Flix): (Type, Type) = {
+    // Make the context's level available
+    // This is a def rather than a val because c is mutable.
+    implicit def level: Level = c.getLevel
```
If we call c.getLevel then we have to call it every time we create a var, which is tedious.

An implicit def is the same as an implicit val, it's just not evaluated eagerly. That means that if `c.level` changes, a `def` will reflect that change, whereas a `val` will not.

mlutze replied:
I'd like to remove levels anyway after this.

magnus-madsen replied:
Jonathans suggestion or just leave this as-is, until a refactor. (Update ticket).

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+      case Expr.Var(sym, loc) => (sym.tvar, Type.Pure)
+      case Expr.Def(sym, tvar, loc) =>
+        val defn = root.defs(sym)
+        val (tconstrs, defTpe) = Scheme.instantiate(defn.spec.sc, loc.asSynthetic)
```
`cconstrs`?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val defn = root.defs(sym)
+        val (tconstrs, defTpe) = Scheme.instantiate(defn.spec.sc, loc.asSynthetic)
+        c.unifyTypeM(tvar, defTpe, loc)
+        c.addClassConstraintsM(tconstrs, loc)
```
Yes

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+            (resTpe, resEff)
+
+          case Ast.CheckedCastType.EffectCast =>
+            // We simply union the purity and effect with a fresh variable.
```
Outdated

magnus-madsen replied:
Can we update both comments to make them use the same wording?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.CheckedCast(cast, exp, tvar, evar, loc) =>
```
Add a comment that this is sound (see the other expressions).

JonathanStarup replied:
Ast.CheckedCastType.TypeCast looks unsound to me, is it checked post-typechecking?

mlutze replied:
Yes, in the Safety phase

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        }
+
+      case Expr.UncheckedCast(exp, declaredTpe, declaredEff, tvar, loc) =>
+        // A cast expression is unsound; the type system assumes the declared type is correct.
```
declared type and effect?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = declaredEff.getOrElse(actualEff)
+        (resTpe, resEff)
+
+      case Expr.UncheckedMaskingCast(exp, loc) =>
```
I dont remember what this does. Can we get a comment like the others.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+      case Expr.Without(exp, effUse, loc) =>
+        val effType = Type.Cst(TypeConstructor.Effect(effUse.sym), effUse.loc)
+        //        val expected = Type.mkDifference(Type.freshVar(Kind.Bool, loc), effType, loc)
+        // TODO EFF-MIGRATION use expected
```
Drop this stuff, but add a short comment about what to do once sets arrive.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = eff
+        (resTpe, resEff)
+
+      case Expr.TryCatch(exp, rules, loc) =>
```
Lets add a visitTryCatchRule

JonathanStarup replied:
could also add `visitRules` a la my suggestion in match so the main snippet here is more succinct

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val ops = effect.ops.map(op => op.sym -> op).toMap
+        val continuationEffect = Type.freshVar(Kind.Eff, loc)
+
+        def unifyFormalParams(op: Symbol.OpSym, expected: List[KindedAst.FormalParam], actual: List[KindedAst.FormalParam]): Unit = {
```
move out

magnus-madsen replied:
and document

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+          c.expectTypeArguments(op, expectedTypes = expected.map(_.tpe), actualTypes = actual.map(_.tpe), actual.map(_.loc), loc)
+        }
+
+        def visitHandlerRule(rule: KindedAst.HandlerRule, tryBlockTpe: Type): (Type, Type) = rule match {
```
move out and document

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = Type.mkUnion(eff :: effs, loc)
+        (resTpe, resEff)
+
+      case Expr.TryWith(exp, effUse, rules, tvar, loc) =>
```
Can we clean this code up? At least a bit and add some useful comments? :)

magnus-madsen commented:
Important: Lets immediately start calling this code from somewhere, even if we throw away the result!

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val (tpes, effs) = exps.map(visitExp).unzip
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(expected = regionType, actual = tpe, exp.loc)
+        val elmTpe = c.unifyAllTypesM(tpes, Kind.Star, loc)
```
We can introduce a variable and add it to the unified types in the case where it's needed.

magnus-madsen replied:
I don't know. I am just observing an inconsistency.

mlutze replied:
listed

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resultEff = Type.freshVar(Kind.Eff, loc) // TODO ASSOC-TYPES should be continuationEffect
+        (resultTpe, resultEff)
+
+      case Expr.Do(op, args, tvar, loc) =>
```
args -> exps

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        (resultTpe, resultEff)
+
+      case Expr.Do(op, args, tvar, loc) =>
+        val effect = root.effects(op.sym.eff)
```
effect -> eff or perhaps decl?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+
+      case Expr.Do(op, args, tvar, loc) =>
+        val effect = root.effects(op.sym.eff)
+        val operation = effect.ops.find(_.sym == op.sym)
```
operation -> op

magnus-madsen replied:
extract this line into a function.

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+            // Case 2: Lookup the sym and instantiate its scheme.
+            val sig = root.classes(sym.clazz).sigs(sym)
+            val (tconstrs1, declaredType) = Scheme.instantiate(sig.spec.sc, loc2.asSynthetic)
+            val constrs1 = tconstrs1.map(_.copy(loc = loc))
```
likely bug

But let's handle it in a followup. I'd like to avoid semantic changes if possible.

magnus-madsen replied:
Ooookay? :)

mlutze replied:
I also have a few changes on top of this PR and I don't want to get buried in conflicts.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val (_, _) = args.map(visitExp).unzip
+        val resTpe = classTpe
+        val resEff = Type.IO
+        (resTpe, resEff)
```
BIG TEXT: I would think that every case would include calls to c.unify unless there is a good reason (e.g. constant). 

Could we double check that? If there is no such call it should be documented imho! "This expression does not generate constraints because ...".

Presumably getFlixType does not generate constraints!

Do we not need to match the arguments against the types of the constructor? Am I stupid?

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val (defaultType, eff2) = visitDefaultRule(default)
+        c.unifyAllTypesM(tvar :: defaultType :: ruleTypes, Kind.Star, loc)
+        val resTpe = tvar
+        val resEff = Type.mkUnion(regionVar :: eff2 :: ruleEffs, loc)
```
Maybe at the end of the day, you can do a search for `::`.

mlutze replied:
Ditto for N things

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = Type.mkUnion(regionVar :: eff2 :: ruleEffs, loc)
+        (resTpe, resEff)
+
+      case Expr.Spawn(exp1, exp2, loc) =>
```
We need a comment here that we don't fully understand yet what the type of spawn should be.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+      case e@Expr.FixpointSolve(exp, loc) => SchemaConstraintGeneration.visitFixpointSolve(e)
+      case e@Expr.FixpointFilter(pred, exp, tvar, loc) => SchemaConstraintGeneration.visitFixpointFilter(e)
+      case e@Expr.FixpointInject(exp, pred, tvar, loc) => SchemaConstraintGeneration.visitFixpointInject(e)
+      case e@Expr.FixpointProject(pred, exp1, exp2, tvar, loc) => SchemaConstraintGeneration.visitFixpointProject(e)
```
e: , for all

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+    implicit def level: Level = c.getLevel
+
+    pat0 match {
+
```
linebreak

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+    */
+  def visitPattern(pat0: KindedAst.Pattern)(implicit c: TypeContext, root: KindedAst.Root, flix: Flix): Type = {
+    // Make the context's level available
+    // This is a def rather than a val because c is mutable.
```
update comment same as above.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+        // The tag type is a function from the type of variant to the type of the enum.
+        //
+        val tpe = visitPattern(pat)
+        c.unifyTypeM(tagType, Type.mkPureArrow(tpe, tvar, loc), loc)
```
simplify as above.

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+        val freshRowVar = Type.freshVar(Kind.RecordRow, loc.asSynthetic)
+        val freshRecord = Type.mkRecord(freshRowVar, loc.asSynthetic)
+
+        def mkRecordType(patTypes: List[(Name.Label, Type, SourceLocation)]): Type = {
```
extract

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+  }
+
+  /**
+    * Generates constraints for the record label pattern.
```
What does this return?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        // effect type is AT LEAST the inner expression's effect
+        val atLeastEff = Type.mkUnion(eff, Type.freshVar(Kind.Eff, loc.asSynthetic), loc.asSynthetic)
+        c.unifyTypeM(atLeastEff, evar, loc)
+        // result type is whatever is needed for the hole
```
This is for hole-with-expression, which is different, I guess?

magnus-madsen replied:
Thanks for the info. I suggest we make this just like hole. Then, once we have sets and subeffecting, we get the right solution.

JonathanStarup replied:
I think the current code is the correct type rule. if you think of `???(???, someExpression: t \ ef, ???)` (where `???` has unknown effect) then it will have any type and effect `ef + alpha`

mlutze replied:
listed

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+
+        val (ruleTypes, ruleEffs) = rules.map(visitSelectRule).unzip
+        val (defaultType, eff2) = visitDefaultRule(default)
+        c.unifyAllTypesM(tvar :: defaultType :: ruleTypes, Kind.Star, loc)
```
`ruleTypes` is a list so it's N things unified, not just 3

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+        (resTpe, resEff)
+
+      case KindedAst.Expr.Unary(sop, exp, tvar, loc) => sop match {
+        case SemanticOp.BoolOp.Not =>
```
It seems all of these unary operations could use the same `helper(exp, tvar, type)` function

mlutze replied:
listed

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+            fparams.foreach(visitFormalParam)
+            val (bodyTpe, bodyEff) = visitExp(exp)
+            c.expectTypeM(expected = returnTpe, actual = bodyTpe, exp.loc)
+          // TODO ASSOC-TYPES check eff matches declared eff ?
```
A probable bug. To be addressed properly in followup.

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+
+      case KindedAst.Expr.Binary(sop, exp1, exp2, tvar, loc) => sop match {
+
+        case SemanticOp.BoolOp.And | SemanticOp.BoolOp.Or =>
```
These could also all share a helper function, then itd be easier to read IMO, the reader just have to verify the hardcoded types of each operation, not all the copied code.

mlutze replied:
listed

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+
+      case Expr.Let(sym, _, exp1, exp2, loc) =>
+        val (tpe1, eff1) = visitExp(exp1)
+        c.unifyTypeM(sym.tvar, tpe1, loc)
```
let uses `loc` and letrec uses `exp1.loc`. which is best?

mlutze replied:
exp1.loc is more precise. fixed

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        c.expectTypeM(expected = arrayType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = Type.Int32, actual = tpe2, exp2.loc)
+        c.expectTypeM(expected = elmVar, actual = tpe3, exp3.loc)
+        c.unifyTypeM(evar, Type.mkUnion(List(regionVar, eff1, eff2, eff3), loc), loc)
```
Adding to followup list

JonathanStarup replied:
could also be varargs no? `mkUnion(tpe1: Type, tpes: Type*): Type`

mlutze replied:
needs a location though

JonathanStarup replied:
could be `mkUnion(loc, tpe, tpes)` but not as cool no

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+        (resTpe, resEff)
+
+      case Expr.LetRec(sym, _, _, exp1, exp2, loc) =>
+        // exp1 is known to be a lambda syntactically
```
I don't think that matters to the rule? if not then we should just remove the comment.

mlutze replied:
@magnus-madsen Fight Jonathan

magnus-madsen replied:
I think the comment is there to remind us not to check that it is a lambda?

JonathanStarup replied:
there is no wellformedness checking other places in constraint gen. It also doesn't matter to the type rule whether or not its a lambda.

mlutze replied:
opting for no further action

open a PR if you disagree

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.Scope(sym, regionVar, exp, evar, loc) =>
```
Maybe `enterRegion` should be a higher order function instead of separate enter/exit calls?

mlutze replied:
It must be inside because it must be resolved before purification happens.

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val resEff = evar
+        (resTpe, resEff)
+
+      case Expr.Match(exp, rules, loc) =>
```
I think you could make a helper function for guards fx. since it justs checks that everything is bool and returns a list of effects. That would help the outer readability

JonathanStarup replied:
similar for visitPatterns, and visitBodies.
```scala
val (tpe, eff) = visitExp(exp)
visitPatterns(rules.map(_.pat), tpe, ...)
val guardEffs = visitGuards(rules.flatMap(_.guard), ...)
val (bodyTpe, bodyEffs) = visitBodies(rules.map(_.exp), ...)
val resTpe = bodyTpe
val resEff = Type.mkUnion(eff :: guardEffs ::: bodyEffs, loc)
```

JonathanStarup replied:
perhaps there can be a helper function for branching expressions. Can be used for if branches, match branches, typematch branches

mlutze replied:
I opted for `visitMatchRule` since it's consistent with what we do in other phases

JonathanStarup reviewed (changes requested):
I got too hungry at match, will do more later

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+        (resTpe, resEff)
+
+      case Expr.RecordEmpty(loc) =>
+        val resTpe = Type.mkRecord(Type.RecordRowEmpty, loc)
```
`Type.RecordRowEmpty` should be `Type.mkRecordRowEmpty(loc)` right? same thing in the next case

mlutze replied:
eh, it creates an extra object for no real gain (type locations for inferred types are evil)

magnus-madsen replied:
I agree with Matt

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+        val (tpe, eff) = visitExp(exp)
+        c.expectTypeM(Type.mkArray(elmVar, regionVar, loc), tpe, exp.loc)
+        val resTpe = Type.Int32
+        val resEff = eff
```
I almost fell into the trap of suggesting a union was missing, maybe a comment is warranted?

mlutze replied:
I think if I add a comment then Magnus will tell me to remove it :P

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+        val (tpe1, eff1) = visitExp(exp1)
+        val (tpe2, eff2) = visitExp(exp2)
+        c.expectTypeM(expected = refType, actual = tpe1, exp1.loc)
+        c.expectTypeM(expected = elmVar, actual = tpe2, exp2.loc)
```
I dont think this one should use expected since elmVar is just a variable

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+
+      case Expr.InstanceOf(exp, _, _) =>
+        val (_, eff) = visitExp(exp)
+        c.expectTypeM(expected = Type.Pure, actual = eff, exp.loc)
```
I don't think this should be there, why would it matter that the expression being checked is pure? if it is something impure you could just letbind it so I'm not sure what it achieves. @magnus-madsen ?

magnus-madsen replied:
Agree. It should not require purity. Good catch!

mlutze replied:
listed

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+        val (actualTyp, actualEff) = visitExp(exp)
+        c.unifyTypeM(tvar, declaredTpe.getOrElse(actualTyp), loc)
+        val resTpe = tvar
+        val resEff = declaredEff.getOrElse(actualEff)
```
This could be nominated to have an `evar` so this getOrElse logic isn't copied in reconstruction.

magnus-madsen replied:
Yes

mlutze replied:
IIRC we wanted to avoid introducing another boolean variable at all costs

mlutze replied:
listed

JonathanStarup commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+        c.unifyTypeM(resTpe, tvar, loc)
+        resTpe
+
+      case KindedAst.Pattern.RecordEmpty(loc) => Type.mkRecord(Type.RecordRowEmpty, loc)
```
Type.mkRecordRowEmpty(loc)

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+            val declaredResultType = declaredType.arrowResultType
+
+            val (tpes, effs) = exps.map(visitExp).unzip
+            c.expectTypeArguments(sym, declaredArgumentTypes, tpes, exps.map(_.loc), loc)
```
per discussion, this problem seems to be addressed in an earlier phase where things are desugared to match the number of args

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1081 @@
+      case Expr.Ascribe(exp, expectedTpe, expectedEff, tvar, loc) =>
+        // An ascribe expression is sound; the type system checks that the declared type matches the inferred type.
+        val (actualTpe, actualEff) = visitExp(exp)
+        c.expectTypeM(expected = expectedTpe.getOrElse(Type.freshVar(Kind.Star, loc)), actual = actualTpe, loc)
```
I thought about that too. But it seems it would be less elegant.

JonathanStarup replied:
I think its more elegant?
```
c.expectTypeM(expected = expectedTpe.getOrElse(Type.freshVar(Kind.Star, loc)), actual = actualTpe, loc)
// vs
expectedTpe.foreach(t => c.expectTypeM(expected = t, actual = actualTpe, loc))
```

magnus-madsen commented:
The reason to be so picky is that we really don't want bugs in the generation. We want all our bugs in the solving, so we can focus our attention there.

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        //
+        // The tag type is a function from the type of variant to the type of the enum.
+        //
+        val (tpe, eff) = visitExp(exp)
```
I prefer to avoid them, as they can get lost over the edge of the screen

magnus-madsen replied:
![CleanShot 2024-02-29 at 21 04 12@2x](https://github.com/flix/flix/assets/2287953/cbd7fd79-b889-4b6a-9f7f-cd17df3a038f)

JonathanStarup replied:
See, struggling to get 80 width of code

![image](https://github.com/flix/flix/assets/32037926/67345d5e-9133-4b92-b00a-bafd96988076)

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1071 @@
+        val sig = root.classes(sym.clazz).sigs(sym)
+        val (tconstrs, sigTpe) = Scheme.instantiate(sig.spec.sc, loc.asSynthetic)
+        c.unifyTypeM(tvar, sigTpe, loc)
+        c.addClassConstraintsM(tconstrs, loc)
```
Maybe. I'll follow up.

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala:
```diff
@@ -0,0 +1,1069 @@
+        // - have better performance (we don't generate unnecessary type variables)
+        //
+        val knownTarget = exp match {
+          case KindedAst.Expr.Def(sym, tvar2, loc2) =>
```
this area mixes with some corrections I made on top of this PR

I'll fix this in the next PR

mlutze replied:
listed

mlutze commented:
I've addressed all the comments. I have deferred a lot of them because when I rebase the constraint resolution PR on top of this one, I am certain I will find errors, and I want to be sure that those are problems with my translation, rather than problems with my trying to fix minor bugs.

I also am already aware of some problems here wrt associated types and effects. But I want to isolate the fixes to those problems.

magnus-madsen commented:
@mlutze Ready for me to do another pass then?

When will we do some of the follow up things?

mlutze commented:
> @mlutze Ready for me to do another pass then?
> 
> When will we do some of the follow up things? Before merging in the solver, after the rebase?

Yes.

My plan is: merge, rebase, fix new bugs if present, rebase, fix associated type bugs, rebase, fix bugs from list, rebase, merge solver, fix style stuff from the list.

Only the first rebase should be painful.

magnus-madsen commented:
@mlutze Merged!
